src/goto-programs/goto_check.cpp

```diff
@@ -212,10 +212,32 @@ void goto_checkt::overflow_check(
   if(is_pointer_type(*expr->get_sub_expr(0)))
     return;
 
+  assert(
+    is_add2t(expr) || is_sub2t(expr) || is_mul2t(expr) || is_shl2t(expr) ||
+    is_neg2t(expr));
+
+  expr2tc overflow, new_expr = expr;
+  expr2tc side_1 = *expr->get_sub_expr(0);
+  expr2tc side_2 = *expr->get_sub_expr(1);
+
+  if(side_1->type != side_2->type)
+  {
+    typecast2tc casted_size_2(side_1->type, side_2);
+    if(is_add2t(expr))
+      new_expr = add2tc(side_1->type, side_1, casted_size_2);
+    else if(is_sub2t(expr))
+      new_expr = sub2tc(side_1->type, side_1, casted_size_2);
+    else if(is_mul2t(expr))
+      new_expr = mul2tc(side_1->type, side_1, casted_size_2);
+    else if(is_shl2t(expr))
+      new_expr = shl2tc(side_1->type, side_1, casted_size_2);
```

Comment:
Instead of `if...else-if...else` I think that using a switch-case with `expr2t::expr_ids` would be better.